cmd/soroban-cli/src/log.rs

```diff
@@ -1,13 +1,57 @@
+use crate::xdr;
 pub mod auth;
 pub mod budget;
+pub mod contract_event;
 pub mod cost;
 pub mod diagnostic_event;
 pub mod footprint;
 pub mod host_event;
+#[allow(clippy::module_name_repetitions)]
+pub mod log_event;
 
 pub use auth::*;
 pub use budget::*;
+pub use contract_event::*;
 pub use cost::*;
 pub use diagnostic_event::*;
 pub use footprint::*;
 pub use host_event::*;
+pub use log_event::*;
+
+pub fn events(events: &[xdr::DiagnosticEvent]) {
+    let (contract_events, other_events): (Vec<_>, Vec<_>) =
+        events.iter().partition(|e| is_contract_event(e));
+    contract_event::contract_events(&contract_events, tracing::Level::INFO);
+    let (log_events, other_events): (Vec<_>, Vec<_>) =
+        other_events.into_iter().partition(|e| is_log_event(e));
+    log_event::log_events(&log_events, tracing::Level::INFO);
+    diagnostic_event::diagnostic_events(&other_events, tracing::Level::DEBUG);
```

Comment:
This logic results in the events being printed out-of-order. All of the contract events print, then all of the log events, then all of the diagnostic events. It's pretty important that we preserve the order and keep them interleaved otherwise diagnostic events that occurred prior to a contract event won't be observed as occurring prior.